s2/region_term_indexer.go

```diff
@@ -0,0 +1,441 @@
+	// minLevel and maxLevel control the minimum and maximum size of the
+	// S2Cells used to approximate regions.  Setting these parameters
+	// appropriately can reduce the size of the index and speed up queries by
+	// reducing the number of terms needed.  For example, if you know that
+	// your query regions will rarely be less than 100 meters in width, then
+	// you could set maxLevel to 100.
+	//
+	// This restricts the index to S2Cells that are approximately 100 meters
+	// across or larger.  Similar, if you know that query regions will rarely
+	// be larger than 1000km across, then you could set minLevel similarly.
+	//
+	// If minLevel is set too high, then large regions may generate too
+	// many query terms.  If maxLevel() set too low, then small query
+	// regions will not be able to discriminate which regions they intersect
+	// very precisely and may return many more candidates than necessary.
+	//
+	// If you have no idea about the scale of the regions being queried,
+	// it is perfectly fine to set minLevel to 0 and maxLevel to 30.
+	// The only drawback is that may result in a larger index and slower queries.
+	//
+	// The default parameter values are suitable for query regions ranging
+	// from about 100 meters to 3000 km across.
+	//
+	// DEFAULT: 4  (average cell width == 600km)
+	minLevel int
+
+	// DEFAULT: 16 (average cell width == 150m)
+	maxLevel int
+
+	// Setting levelMod to a value greater than 1 increases the effective
+	// branching factor of the S2Cell hierarchy by skipping some levels.  For
+	// example, if levelMod to 2 then every second level is skipped (which
+	// increases the effective branching factor to 16).  You might want to
+	// consider doing this if your query regions are typically very small
+	// (e.g., single points) and you don't mind increasing the index size
+	// (since skipping levels will reduce the accuracy of cell coverings for a
+	// given maxCells limit).
+	//
+	// DEFAULT: 1  (don't skip any cell levels)
+	levelMod int
+
+	// If your index will only contain points (rather than regions), be sure
+	// to set this flag.  This will generate smaller and faster queries that
+	// are specialized for the points-only case.
+	//
+	// With the default quality settings, this flag reduces the number of
+	// query terms by about a factor of two.  (The improvement gets smaller
+	// as maxCells is increased, but there is really no reason not to use
+	// this flag if your index consists entirely of points.)
+	//
+	// DEFAULT: false
+	pointsOnly bool
+
+	// If true, the index will be optimized for space rather than for query
+	// time.  With the default quality settings, this flag reduces the number
+	// of index terms and increases the number of query terms by the same
+	// factor (approximately 1.3).  The factor increases up to a limiting
+	// ratio of 2.0 as maxCells is increased.
+	//
+	// CAVEAT: This option has no effect if the index contains only points.
+	//
+	// DEFAULT: false
+	optimizeSpace bool
+}
+
```

Comment:
The fact we need to explicitly set levelMod on every unit test feels like a point of trouble that's going to bite people down the road.  A default options instance should work safely without the user needing to know to set certain fields. 

e.g. Anyone who forgets to set it will get stuck with infinite loops because several methods do this:

for (level - rti.options.levelMod) >= rti.options.minLevel {

I think there should be a DefaultRegionTermIndexerOptions() that gives the caller a safe variable to start from when they need an options.

Reply:
There will still be a foot gun. @alan-strohm @flwyd Thoughts on idiomatic ways to check for valid options?

Reply:
> Do users of the indexer need to be able to inspect the option state (public fields), or would a set-only interface, as in the functional options pattern, suffice?

The accessors are only so the builder and related functions can get the options.  Clients only need to set them.